### PR TITLE
Adding iree/runtime/ as a higher-level C API.

### LINKED ISSUE #5184: Proposed simplification of iree_status_allocate_vf() api
The `iree_status_allocate_vf` function signature is currently 
```C
IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
iree_status_allocate_vf(iree_status_code_t code, const char* file,
                        uint32_t line, const char* format, va_list varargs_0,
                        va_list varargs_1)
```
It has two va_list arguments due to the internal implementation requirements. 
I propose we remove `varargs_1` and simply do a `va_copy` on function entry. I looked at the implementations of `va_start` and `va_copy` and this will actually be less work than the current implementation's creating two `va_lists`. It will also make the public api more sensible for callers.
```C
IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
iree_status_allocate_vf(iree_status_code_t code, const char* file,
                        uint32_t line, const char* format, va_list varargs_0)
  // make a copy so we can walk the arguments once to determine storage size
  // and once to print the values.
  va_list varargs_1;
  va_copy(varargs_1, varargs_0);
```
I have already implemented this and it seems to work correctly.

### LINKED ISSUE #5631: Add a spot for user-friendly utils to replace vm_util and make use easier.
Thinking something along the lines of `iree_runtime_t` > `iree_runtime_process_t` > `iree_runtime_invocation_t` > `iree_runtime_invocation_builder_t`.

These would wrap up lifetime management, error checking, and add utility functions that interact with the HAL with abandon (buffer view copy/clone/placement/etc).

### PR DESCRIPTION
**Pull request by benvanik**
This wraps up a lot of the state required to use the VM/HAL and makes
error handling much simpler as there are fewer allocations to track and
errors to check. It'll also act as a good place for standard file format
handling that we want to be part of core.

This initial version is just enough to get going. Future versions will add
more `iree_runtime_call_*` helpers to make constructing calls easier.

Note that the tradeoff of using this higher level interface is that there
are more implicit dependencies and less opportunities for LTO to strip
things. The lower level APIs are still the best route for deployments where
binary size or platform compatibility is a primary concern.

This will be used to replace vm_util in all of the IREE tools once SIP
and the signature parser are removed.

Fixes #5631.
Fixes #5184.
Progress on #5238.
Progress on #3175.

### CODE REVIEW

**Comment by benvanik** [comment]
This first PR is mostly just to get something in place and set up the skeleton hello world demo. Very open to naming suggestions. Once this settles for a bit I'll be going back and switching over most of our tools/vm_util users to this.

File: iree/runtime/demo/hello_world_explained.c [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+          IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+          // What access to allow to this memory (this is .rodata so READ only):
+          IREE_HAL_MEMORY_ACCESS_READ,
+          // Intended usage of the buffer (transfers, dispatches, etc):
+          IREE_HAL_BUFFER_USAGE_ALL,
+          // The actual heap buffer to wrap or clone and its allocator:
+          iree_make_byte_span((void*)arg0_data, sizeof(arg0_data)),
+          iree_allocator_null(),
+          // Buffer view + storage are returned and owned by the caller:
+          &arg0);
```
Comment (ScottTodd) [inline iree/runtime/demo/hello_world_explained.c:216]: This is still pretty verbose for a higher level API. The comments help, but I'm still wondering in what cases / how frequently users will want to actually change the memory type, memory access, and buffer usage. Could we add a helper that uses these as defaults?

File: iree/runtime/session.h [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+// Returns the instance the session uses for shared resources.
+IREE_API_EXPORT iree_runtime_instance_t* iree_runtime_session_instance(
+    const iree_runtime_session_t* session);
+
+// Returns the VM context used to load and link modules.
+// The context can be used to perform additional reflection over the loaded
+// modules or load additional modules (if supported).
+IREE_API_EXPORT iree_vm_context_t* iree_runtime_session_context(
+    const iree_runtime_session_t* session);
```
Comment (ScottTodd) [inline iree/runtime/session.h:118]: Ok, I was wondering about this.

The higher level API is built on top of the low level API, but still offers access to the underlying structures. So it's possible to use the high level API to get set up then use the low level structures for full expressivity?

File: iree/runtime/session.h [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+// NOTE: this device allocator will not be available until initialized by a
+// user module and will return NULL if queried prior.
+IREE_API_EXPORT iree_hal_allocator_t* iree_runtime_session_device_allocator(
+    const iree_runtime_session_t* session);
+
+// Appends the given |module| to the context.
+// The module will be retained by the context.
+//
+// NOTE: only valid if the context is not yet frozen; see
+// iree_runtime_session_create for more information.
```
Comment (ScottTodd) [inline iree/runtime/session.h:140]: I'm not quite following these NOTEs. I see `iree_runtime_session_create_with_device`, but not `iree_runtime_session_create`, and no mentions of "frozen" contexts?

File: iree/runtime/session.h [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+// of |full_name| or returns IREE_STATUS_NOT_FOUND. The function reference is
+// valid for the lifetime of |session|.
+//
+// The function name matches the original MLIR module and function symbols.
+// Example:
+//   module @foo {
+//     func @bar()
+//   }
+// The full name of '@bar' is 'foo.bar'.
+// By default modules have the name 'module'.
```
Comment (ScottTodd) [inline iree/runtime/session.h:179]: How can module names be changed? I've only ever seen 'module'...

I think Ed was asking about this also. It would be nice to have a sample that loads two different modules to show that there aren't name conflicts o_o

File: iree/runtime/session.h [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+IREE_API_EXPORT iree_status_t iree_runtime_session_lookup_function(
+    const iree_runtime_session_t* session, iree_string_view_t full_name,
+    iree_vm_function_t* out_function);
+
+// Synchronously issues a generic function call.
```
Comment (ScottTodd) [inline iree/runtime/session.h:184]: Do you see asynchronous calls being supported by iree/runtime/? Or would that be an exclusively low level API?

File: iree/runtime/call.h [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    iree_runtime_call_t* call, iree_runtime_call_flags_t flags);
+
+//===----------------------------------------------------------------------===//
+// Helpers for defining call I/O
+//===----------------------------------------------------------------------===//
+// NOTE: these are mostly useful for one-shot tests and samples. Applications
+// that will be reusing the same inputs and outputs should prefer to track them
+// themselves. If applications are able it's strongly recommended that it
+// produces and consumes the iree_hal_buffer_ts directly to avoid additional
+// copies and allocations.
```
Comment (ScottTodd) [inline iree/runtime/call.h:108]: nit: grammar (choose singular or plural, don't mix)

```suggestion
// themselves. If applications are able it's strongly recommended that they
// produce and consume the iree_hal_buffer_ts directly to avoid additional
// copies and allocations.
```

**Review by ScottTodd — APPROVED** [review]
Cool!

General impressions:

* This looks significantly easier to use, particularly for routing data into and out of function calls.
* Having two C APIs concerns me a bit, and I'm now wondering which (low level? high level? both?) would make sense to wrap in language bindings. They aren't _that_ different though, and it looks like you can mix them by getting the low level structs from the higher level structs.
* Some of the details around multi-tenant uses seem a bit fuzzy to me. The documentation in the code is clear, but without some practical uses actually exercising that code and showing the separation I'm skeptical.

File: iree/vm/context.h [hunk 7]
```diff
@@ -69,7 +69,7 @@ iree_vm_context_state_resolver(const iree_vm_context_t* context);
 
 // Sets |out_module_state| to the context-specific state for the given |module|.
 // The state is owned by the context and will only be live for as long as the
-// context is.
+// context does.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/vm/context.h:72]: lol misread

File: iree/modules/hal/hal_module.h [hunk 8]
```diff
@@ -54,6 +54,22 @@ IREE_API_EXPORT iree_status_t
 iree_hal_module_create(iree_hal_device_t* device, iree_allocator_t allocator,
                        iree_vm_module_t** out_module);
 
+// Returns the device currently in use by the HAL module.
+// Returns NULL if no device has been initialized yet.
+IREE_API_EXPORT iree_hal_device_t* iree_hal_module_state_device(
+    iree_vm_module_state_t* module_state);
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/modules/hal/hal_module.h:60]: This is forward-looking to #5724 where you won't have to provide a device and instead the user module will select the device itself. That's the other work I have in-flight. You'll still be able to manually specify devices (which is useful to avoid enumeration when you are targeting a single device), but both paths will be valid.

File: iree/runtime/demo/hello_world_explained.c (hunk 1 above)
Comment (benvanik) [reply iree/runtime/demo/hello_world_explained.c:216]: Yeah it's a lot of characters, but easy to copy and intellisense works really well. This isn't a higher-level API - it's the HAL API. And if there's one place to have verbosity when using IREE it's for buffer management - buffers are kind of important :)

There are no defaults that are good for this kind of stuff - you _must_ know whether you want your memory to live on the host or the device and you _must_ know if you want it read-only or read-write. What's cheap on the local CPU device or GPU with unified memory can be catastrophically slow on a sandbox/enclave or functionally impossible on a GPU with discrete memory (see our out-of-memories with HOST_LOCAL|DEVICE_VISIBLE memory on AMD after only 256MB of allocations). The alternative is to make multiple functions with names like `iree_hal_buffer_view_wrap_or_clone_heap_buffer_on_device_and_readonly`, which is just API bloat and makes changing things in the future harder. This way, new enums can be added without needing to plumb through new API calls.

Buffer usage could _maybe_ be defaulted to ALL, however once you have the other two I don't see a reason to skip it - and IREE_HAL_BUFFER_USAGE_CONSTANT can be extremely useful (indicates that the memory will never change out from under the HAL so when mirrored on to a device/remote host/sandbox it's safe to assume it'll never change, etc). If we add more flags here (like EXPORTABLE or ENCYRPTED or whatever) we don't want to have to plumb through new methods for taht.

At the point you want something higher level than this you move out of the realm of things we care about in core and into binding land and I'd rather have this API be something bindings could use with full fidelity. Trying to hand-hold here just leads to cases where people have to choose between using the nicer API _or_ having access to that one lower-level flag that lets them do what they need. I'm trying not to force that tradeoff :)

Basically: if you are using C you can deal with 3 extra arguments that are load-bearing to correctness and performance. If you don't care, you can use something like the tflite API - and that's a pretty good example of the mess I'm trying to avoid here (see EnsureTensorDataIsReadable, GetBufferHandle, SetAllowBufferHandleOutput, SetCustomAllocationForTensor, etc).

File: iree/runtime/session.h (hunk 2 above)
Comment (benvanik) [reply iree/runtime/session.h:118]: Yep! So registering custom modules is just getting the context and calling register_module on it without the need for additional plumbing here:
```c
iree_vm_context_register_modules(iree_runtime_session_context(session), ...);
```

Avoids a lot of duplication and potential for errors to be introduced in the plumbing.

File: iree/runtime/session.h (hunk 4 above)
Comment (benvanik) [reply iree/runtime/session.h:179]:
```
module @foo {
func @bar() ...
}
```

By default modules are called 'module' but you can just spell them out if you want. This is a quirk of the MLIR parser - it always inserts a default module wrapper if you omit one. But that's mostly an issue for the importers: python may name the modules after the parent class, etc.

File: iree/runtime/session.h (hunk 5 above)
Comment (benvanik) [reply iree/runtime/session.h:184]: Yep! Probably via `iree_runtime_call_t` (or an async_ variant of it) - that way the user would have a handle to state they could use to query/cancel/etc. I wanted to avoid that here until the ABI refresh is done and we can plumb through semaphores again.

File: iree/runtime/session.h (hunk 3 above)
Comment (benvanik) [reply iree/runtime/session.h:140]: ah yeah, I removed that method until #5724 (as today you have to specify the device) - will fix the note. Freezing contexts is also something not yet implemented so I'll file an issue (today contexts are frozen by default when specifying modules during creation and a manual way to freeze them needs to be added).

File: iree/runtime/demo/hello_world_explained.c (hunk 1 above)
Comment (benvanik) [reply iree/runtime/demo/hello_world_explained.c:216]: (I'm not sure the best way to say all that in the docs/comments, tried not to expound too much in here as I can tend to type too much and I didn't want it to be confusing, but I'm personally a fan of this kind of guidance so open to suggestions on how to best integrate it :)
